<a href="https://colab.research.google.com/github/avinashkumar18/Lab/blob/main/MLT_Lab/20MAI0035_MLT_LAB_FAT.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

[Github link](https://https://github.com/avinashkumar18/Lab/blob/main/MLT_Lab/20MAI0035_MLT_LAB_FAT.ipynb)

In [ ]:
# Importing necessary packages
import pandas as pd
import numpy as np

In [ ]:
# Downloaded student performance dataset from https://archive.ics.uci.edu/ml/index.php
# uploaded into my own github repository and used as my dataset
url = "https://raw.githubusercontent.com/avinashkumar18/Lab/main/MLT_Lab/student-mat.csv"
df = pd.read_csv(url)

In [ ]:
# overview of dataset
df.head()

school sex   age address famsize  ... absences    G1    G2    G3 Unnamed: 33
0     GP   F  18.0       U     GT3  ...      6.0   5.0   6.0   6.0         NaN
1     GP   F  17.0       U     GT3  ...      4.0   5.0   5.0   6.0         NaN
2     GP   F  15.0       U     LE3  ...     10.0   7.0   8.0  10.0         NaN
3     GP   F  15.0       U     GT3  ...      2.0  15.0  14.0  15.0         NaN
4     GP   F  16.0       U     GT3  ...      4.0   6.0  10.0  10.0         NaN

[5 rows x 34 columns]

In [ ]:
# Dimensions of the dataset
df.shape

(396, 34)

In [ ]:
# Info about the dataset
df.describe

<bound method NDFrame.describe of     school  sex   age address famsize  ... absences    G1    G2    G3 Unnamed: 33
0       GP    F  18.0       U     GT3  ...      6.0   5.0   6.0   6.0         NaN
1       GP    F  17.0       U     GT3  ...      4.0   5.0   5.0   6.0         NaN
2       GP    F  15.0       U     LE3  ...     10.0   7.0   8.0  10.0         NaN
3       GP    F  15.0       U     GT3  ...      2.0  15.0  14.0  15.0         NaN
4       GP    F  16.0       U     GT3  ...      4.0   6.0  10.0  10.0         NaN
..     ...  ...   ...     ...     ...  ...      ...   ...   ...   ...         ...
391     MS    M  17.0       U     LE3  ...      3.0  14.0  16.0  16.0         NaN
392     MS    M  21.0       R     GT3  ...      3.0  10.0   8.0   7.0         NaN
393     MS    M  18.0       R     LE3  ...      0.0  11.0  12.0  10.0         NaN
394     MS    M  19.0       U     LE3  ...      5.0   8.0   9.0   9.0         NaN
395    NaN  NaN   NaN     NaN     NaN  ...      NaN   NaN   NaN 

Considering Higher as the target value because We are going to predict if the student can able to go for higher studies based on their current performances

In [ ]:
# Considering Higher as the target value because
# We are going to predict if the student can able
# to go for higher studies based on their current performances
taget = df['higher']
df = df.drop('higher', axis=1)

In [ ]:
# One-hot encoding techinique is done to make free of categorical values
df = pd.get_dummies(df)
df.head()

age  Medu  Fedu  ...  internet_yes  romantic_no  romantic_yes
0  18.0   4.0   4.0  ...             0            1             0
1  17.0   1.0   1.0  ...             1            1             0
2  15.0   1.0   1.0  ...             1            1             0
3  15.0   4.0   2.0  ...             1            0             1
4  16.0   3.0   3.0  ...             0            1             0

[5 rows x 58 columns]

In [ ]:
taget.head()

0    yes
1    yes
2    yes
3    yes
4    yes
Name: higher, dtype: object

In [ ]:
# Encoding the target value with either 1 or 0
target = taget.map({'yes':1,'no':0})

In [ ]:
target.value_counts()

1.0    375
0.0     20
Name: higher, dtype: int64

In [ ]:
# checking if a row contain any NA value
df.isna().sum()

age                    1
Medu                   1
Fedu                   1
traveltime             1
studytime              1
failures               1
famrel                 1
freetime               1
goout                  1
Dalc                   1
Walc                   1
health                 1
absences               1
G1                     1
G2                     1
G3                     1
Unnamed: 33          396
school_GP              0
school_MS              0
sex_F                  0
sex_M                  0
address_R              0
address_U              0
famsize_GT3            0
famsize_LE3            0
Pstatus_A              0
Pstatus_T              0
Mjob_at_home           0
Mjob_health            0
Mjob_other             0
Mjob_services          0
Mjob_teacher           0
Fjob_at_home           0
Fjob_health            0
Fjob_other             0
Fjob_services          0
Fjob_teacher           0
reason_course          0
reason_home            0
reason_other           0


In [ ]:
# The below column is dropped because this column contain only nan values
df = df.drop(['Unnamed: 33'], axis=1)

In [ ]:
#The last row also contains NA values, so they are also dropped
df = df.drop([395])
target = target.drop([395])

In [ ]:
df.tail()

age  Medu  Fedu  ...  internet_yes  romantic_no  romantic_yes
390  20.0   2.0   2.0  ...             0            1             0
391  17.0   3.0   1.0  ...             1            1             0
392  21.0   1.0   1.0  ...             0            1             0
393  18.0   3.0   2.0  ...             1            1             0
394  19.0   1.0   1.0  ...             1            1             0

[5 rows x 57 columns]

In [ ]:
# Naming the column
target.column=['higher']

In [ ]:
# Converting datatype of target similar to input datatype
target= target.astype("category")

In [ ]:
# Assigning the dependent and independent values for model training and pred
x = df
y = target

In [ ]:
# Importing train test split and splitting the dataset
# 30% goes as test data
from sklearn .model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(x, y, test_size=0.3, random_state = 25)

In [ ]:
# Since it does not have binary values, we are normalizing the values
X_binarised_train = X_train.apply(pd.cut, bins=2, labels=[1,0]) 
X_binarised_test = X_test.apply(pd.cut, bins=2, labels=[1,0])

In [ ]:
# Assigning the binary values
X_binarised_train = X_binarised_train.values
X_binarised_test = X_binarised_test.values

In [36]:
# IMporting random forest classifier
from sklearn.ensemble import RandomForestClassifier
rc = RandomForestClassifier()

I used Random forest because, for this moderate small data, a decision tree can give the best result considering all binary values. The time taken is average compared to KNN and SVM

In [37]:
# fitting the model
rc.fit(X_train, Y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [38]:
predictions = rc.predict(X_test)

In [39]:
# confusion matrix
from sklearn.metrics import confusion_matrix
RF_confusion_matrix = confusion_matrix(Y_test, predictions)
RF_confusion_matrix

array([[  0,   7],
       [  0, 112]])

In [40]:
# Accuracy score
from sklearn.metrics import accuracy_score
print(accuracy_score(Y_test, predictions))

0.9411764705882353


Although my model predicted an accuracyy of 94%, I am curious what other models would predict.